src/order/closure.lean

```diff
@@ -118,6 +118,48 @@ c.closure_top
 lemma closure_le_closed_iff_le {x y : α} (hy : c.closed y) : x ≤ y ↔ c x ≤ y :=
 by rw [← c.closure_eq_self_of_mem_closed hy, le_closure_iff]
 
+def mk₂ (f : α → α) (p : α → Prop) (hf : ∀ x, x ≤ f x) (hfp : ∀ x, p (f x))
+  (hmin : ∀ ⦃x y⦄, x ≤ y → p y → f x ≤ y) :
+  closure_operator α :=
+{ to_fun := f,
+  monotone' := λ x y hxy, hmin (le_trans hxy (hf y)) (hfp y),
+  le_closure' := hf,
+  idempotent' := λ x, le_antisymm (hmin (le_refl _) (hfp x)) (hf _) }
+
+lemma mk₂.closed_of_prop {f : α → α} {p : α → Prop} (hf : ∀ x, x ≤ f x) (hfp : ∀ x, p (f x))
+  (hmin : ∀ ⦃x y⦄, x ≤ y → p y → f x ≤ y) {x : α} (hx : p x) :
+  (mk₂ f p hf hfp hmin).closed x :=
+le_antisymm (hmin (le_refl _) hx) (hf _)
+
+def mk₃ (f : α → α) (hf : ∀ x, x ≤ f x) (hmin : ∀ ⦃x y⦄, x ≤ f y → f x ≤ f y) :
+  closure_operator α :=
+{ to_fun := f,
+  monotone' := λ x y hxy, hmin (le_trans hxy (hf y)),
+  le_closure' := hf,
+  idempotent' := λ x, le_antisymm (hmin (le_refl _)) (hf _) }
+
+lemma closure_closure_sup_left {α : Type u} [semilattice_sup α] (c : closure_operator α)
+  (x y : α) :
+  c (c x ⊔ y) = c (x ⊔ y) :=
+le_antisymm ((closure_le_closed_iff_le c (closure_is_closed _ _)).1 (sup_le (c.monotone le_sup_left)
+  (le_trans le_sup_right (le_closure _ _)))) (c.monotone (sup_le_sup_right (le_closure _ _) _))
+
+lemma closure_closure_sup_right {α : Type u} [semilattice_sup α] (c : closure_operator α)
+  (x y : α) :
+  c (x ⊔ c y) = c (x ⊔ y) :=
+by { rw [sup_comm, closure_closure_sup_left, sup_comm] }
+
+lemma closure_closure_sup {α : Type u} [semilattice_sup α] (c : closure_operator α)
+  (x y : α) :
+  c (c x ⊔ c y) = c (x ⊔ y) :=
+by { rw [closure_closure_sup_left, closure_closure_sup_right] }
+
+lemma closure_closure_Sup {α : Type u} {ι : Type v} [complete_lattice α] (c : closure_operator α)
+  (x : ι → α) :
+  c (⨆ i, c (x i)) = c (⨆ i, x i) :=
```

Comment:
If you think the `bsupr` version is useful too, feel free to add it.

Reply:
How easy is it to translate from one to the other?

Reply:
`⨆ i ∈ s, f i` is shorthand for `⨆ i, ⨆ (h : i ∈ s), f i`. Usually you can prove the `bsupr` version by applying the lemma twice, but I struggle to see how to do that here.
I guess you want a lemma to turn `⨆ (h : i ∈ s), c (x i) = c (⨆ (h : i ∈ s))`, which is true as long as `c ⊥ = ⊥`.

Reply:
Actually, `c ⊥ = ⊥` doesn't hold in general (`c` can send everything on `⊤` for example). So I will just keep both versions.
And `closure_supr_Prop` is even wronger. For example, the union of two convex sets if not convex in general.

Reply:
`closure_supr_Prop` doesn't state that though - it's talking about the supr over all proofs, of which there is either one or zero.